kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
+-- * When the `handler` function does not return a true value for a batch, it is retried for up to
+--   `max_retry_time` seconds before the batch is deleted and an error is logged.  Retries are organized
+--   by the queue library using an exponential back-off algorithm with the maximum time between retries
+--   set to `max_retry_delay` seconds.
+
+local workspaces = require "kong.workspaces"
+local semaphore = require "ngx.semaphore"
+
+local DEBUG = ngx.DEBUG
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+local ERR = ngx.ERR
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
+end
+
+
+local Queue = {
+  CAPACITY_WARNING_THRESHOLD = 0.8, -- Threshold to warn that the queue capacity limit is reached
+}
+
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+local function make_queue_key(name)
+  return string.format("%s.%s", workspaces.get_workspace_id(), name)
+end
+
+
+local queues = setmetatable({}, {
+  __newindex = function (self, name, queue)
+    return rawset(self, make_queue_key(name), queue)
+  end,
+  __index = function (self, name)
+    return rawget(self, make_queue_key(name))
+  end
+})
+
+
+function Queue.exists(name)
+  return queues[name] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+local function get_or_create_queue(queue_conf, handler, handler_conf)
+
+  local name = queue_conf.name
+
+  local queue = queues[name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    -- We always use the latest configuration that we have seen for a queue and handler.
+    queue.handler_conf = handler_conf
+    return queue
+  end
+
+  queue = {
+    -- Queue parameters from the enqueue call
+    name = name,
+    handler = handler,
+    handler_conf = handler_conf,
+
+    -- semaphore to count the number of items on the queue and synchronize between enqueue and handler
+    semaphore = semaphore.new(),
+
+    -- `bytes_queued` holds the number of bytes on the queue.  It will be used only if string_capacity is set.
+    bytes_queued = 0,
+    -- `entries` holds the actual queue items.
+    entries = {},
+    -- Pointers into the table that holds the actual queued entries.  `front` points to the oldest, `back` points to
+    -- the newest entry.
+    front = 1,
+    back = 1,
+  }
+  for option, value in pairs(queue_conf) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_at(name, 0, function(_, q)
```

Comment:
And what if the queue is shutting down when this timer is called? Is there a reason that is being ignored? Should we document the reason in the code (since it is not obvious)?

Reply:
> And what if the queue is shutting down when this timer is called? Is there a reason that is being ignored? Should we document the reason in the code (since it is not obvious)?

I don't understand what you mean by "the queue is shutting down".  A queue is shut down when no entries remain on it, and at that point, it is deleted from the `queues` array.  There is no `yield` between the test for queue emptiness, the deletion of the queue from `queues` and the return from the timer handler, so from the perspective of other Lua code, the queue either exists or it does not.

Reply:
>  There is no yield between the test for queue emptiness, the deletion of the queue from queues and the return from the timer handler, so from the perspective of other Lua code, the queue either exists or it does not.

The handler code can do an HTTP request which can and will yield.

Reply:
@hbagdi With the new implementation, the `premature` argument will not normally be set (i.e. it would require a `SIGQUIT` to arrive between the invocation of `timer.named_at` and the invocation of the timer handler).  At this point, we are not making the `premature` flag available to the handler because in all of our use cases, it does not matter whether the queue is drained during normal operation or during shutdown.  I don't think that we want to make that information available to handlers because it'd create a dependency on how queues are implemented and there is no clear use case in which having that information be available in the handler would be beneficial.

Reply:
Marking this as resolved due to no further feedback.